### PR TITLE
Add crash guards & comments for Shields P3A

### PR DESCRIPTION
**Pull request by DJAndries**
Resolves brave/brave-browser#25126

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_shields/browser/brave_shields_p3a.cc [hunk 1]
```diff
@@ -152,16 +161,28 @@ void RecordShieldsDomainSettingCounts(PrefService* profile_prefs,
   const char* below_hg_name = is_fingerprint
                                   ? kDomainFPSettingsBelowHistogramName
                                   : kDomainAdsSettingsBelowHistogramName;
+  // Retrieve a count of domains with a setting above the global setting.
```
Comment (bridiver) [inline components/brave_shields/browser/brave_shields_p3a.cc:164]: It's not clear what "above" means. I think in this context you mean more or less permissive?

File: components/brave_shields/browser/brave_shields_p3a.cc [hunk 2]
```diff
@@ -209,6 +234,13 @@ void MaybeRecordInitialShieldsSettings(PrefService* profile_prefs,
           << " standard=" << ads_counts.standard
           << " agg=" << ads_counts.aggressive;
 
+  // Once the domain setting have been counted via content settings,
+  // update each domain setting count pref with count results.
+  //
+  // These count prefs will be used to keep track of setting counts, and will
+  // be updated via RecordShieldsDomainSettingCountsWithChange whenever a
+  // domain setting is changed. This is more efficient than processing
+  // the content settings upon every change.
```
Comment (DJAndries) [inline components/brave_shields/browser/brave_shields_p3a.cc:243]: @bridiver Here's the "why"

File: components/brave_shields/browser/brave_shields_p3a.cc (hunk 1 above)
Comment (DJAndries) [reply components/brave_shields/browser/brave_shields_p3a.cc:166]: Clarified on lines 105-106. The P3A wiki also describes "above"/"below": https://github.com/brave/brave-browser/wiki/P3A#q51-on-how-many-domains-has-the-user-set-the-adblock-setting-to-be-lower-block-less-than-the-default

File: components/brave_shields/browser/brave_shields_p3a.cc [hunk 3]
```diff
@@ -152,16 +163,28 @@ void RecordShieldsDomainSettingCounts(PrefService* profile_prefs,
...
           << " above=" << above_total << " below=" << below_total;
-  p3a_utils::RecordToHistogramBucket(above_hg_name, kDomainCountBuckets,
-                                     above_total);
-  p3a_utils::RecordToHistogramBucket(below_hg_name, kDomainCountBuckets,
-                                     below_total);
+  if (above_total >= 0) {
+    p3a_utils::RecordToHistogramBucket(above_hg_name, kDomainCountBuckets,
+                                       above_total);
+  } else {
+    LOG(ERROR) << "BraveShieldsP3A: \"domains above global setting\""
```
Comment (bridiver) [inline components/brave_shields/browser/brave_shields_p3a.cc:178]: these should be vlog

Comment (bridiver) [inline components/brave_shields/browser/brave_shields_p3a.cc:178]: I think a DCHECK would be more useful than a log

File: components/brave_shields/browser/brave_shields_p3a.cc [hunk 4]
```diff
@@ -102,20 +102,31 @@ void UpdateDomainSettingCount(PrefService* profile_prefs,
...
                                        ControlType global_setting,
                                        bool count_above) {
   const ControlType* setting_order =
       is_fingerprint ? kFPSettingOrder : kAdsSettingOrder;
+  // Initialized in order to start iteration near the current global_setting
   const ControlType* global_setting_order_it =
       std::find(setting_order, setting_order + kSettingCount, global_setting);
+
   if (global_setting_order_it == setting_order + kSettingCount) {
+    // If the global_setting is unexpectedly not part of the setting_order,
```
Comment (bridiver) [inline components/brave_shields/browser/brave_shields_p3a.cc:118]: if this is unexpected, should it also have a DCHECK?

**Review by bridiver — APPROVED** [review]

**Pull request merged**